tuta-sdk/rust/src/crypto/ecc.rs

```diff
@@ -1,14 +1,101 @@
+use rand_core::CryptoRngCore;
 use zeroize::*;
 
-#[derive(Zeroize, ZeroizeOnDrop)]
-pub struct EccPrivateKey([u8; 32]);
+const ECC_KEY_SIZE: usize = 32;
 
-#[derive(Zeroize, ZeroizeOnDrop)]
+#[derive(ZeroizeOnDrop, Clone)]
+pub struct EccPrivateKey([u8; ECC_KEY_SIZE]);
+
+impl EccPrivateKey {
+    /// Get a reference to the underlying bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_slice()
+    }
+
+    /// Calculate the public key for this private key.
+    pub fn derive_public_key(&self) -> EccPublicKey {
+        use curve25519_dalek::Scalar;
+        use curve25519_dalek::montgomery::MontgomeryPoint;
+
+        // public key = private key * base point
+        let public_key = Zeroizing::new(
+            MontgomeryPoint::mul_base(&*Zeroizing::new(Scalar::from_bytes_mod_order(self.0)))
+        );
+
+        EccPublicKey(public_key.0)
+    }
+
+    /// Attempt to convert a slice of bytes into an ECC key.
+    ///
+    /// Returns `Err` on failure.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, EccKeyError> {
+        match bytes.len() {
+            ECC_KEY_SIZE => Ok(Self(bytes.try_into().unwrap())),
+            actual_size => Err(EccKeyError { actual_size })
+        }
+    }
+
+    fn from_bytes_clamped(bytes: [u8; 32]) -> Self {
+        let clamped = curve25519_dalek::scalar::clamp_integer(bytes);
+        Self(clamped)
+    }
+}
+
+#[derive(ZeroizeOnDrop, Clone)]
 pub struct EccPublicKey([u8; 32]);
 
+#[derive(Clone)]
+pub struct EccKeyPair {
+    pub public_key: EccPublicKey,
+    pub private_key: EccPrivateKey,
+}
+
+impl EccKeyPair {
+    /// Generate a keypair with the given random number generator.
+    pub fn generate<R: CryptoRngCore + ?Sized>(rng: &mut R) -> Self {
+        let mut private_key = [0u8; 32];
+        rng.try_fill_bytes(&mut private_key).unwrap();
```

Comment:
We also do this for the IV and AES Key. Maybe it is worth creating a macro or function to generate a byte array of a specified type?